### PR TITLE
Adding page summarizing modeling changes for v4.0

### PR DESCRIPTION
**Pull request by perrygreenfield**
This is a resubmission of these changes to make the PR simpler.

This PR is primarily to add the contents of changes_for_4.rst in modeling. Though I would also like to see the proper form for the reference to this in CHANGES.rst if someone would helpfully give such an example.

EDIT: Follow up of #8769, #9051

### CODE REVIEW

File: docs/modeling/changes_for_4.rst [hunk 1]
```diff
@@ -1,4 +1,177 @@
...
+The new implementation does away with the metaclass for
+compound models and correspondingly no longer supports
+expressions of model classes, but only expressions of 
+model instances. Previously the expression tree was a private 
+attribute. Now the compound class is itself an expression
+tree. The mapping of constituent parameters to the
+compound is not automatically done except when certain 
+method or attribute calls are made, which are not normally 
+done on subexpressions. There may be some special cases
+where it is necessary to force this by calling the `map_parameters()`
```
Comment (nden) [inline docs/modeling/changes_for_4.rst]:
```suggestion
where it is necessary to force this by calling the `.map_parameters`
```

File: docs/modeling/changes_for_4.rst [hunk 2]
```diff
@@ -1,4 +1,177 @@
...
+values of the parameters were held in an array that the model
+held, and that array held all values for all parameters in one
+array. The drawback of this approach was that compound models
+had to create their own array holding values for all the parameters
+of the compound model. As a result, the parameter values in a 
+compound model now are completely disassociated with those in
+the constituent model that made up the compound model. Changes 
+to one were not reflected in the other, often leadin to confusion.
+
+The new implementation still uses parameter descriptors, but only
```
[the remark this answers is not in the document]
Comment (perrygreenfield) [reply docs/modeling/changes_for_4.rst]: Right. I meant to change that. Thanks for noticing.

File: docs/modeling/changes_for_4.rst [hunk 3]
```diff
@@ -1,4 +1,178 @@
...
+
+        m = m1 + m2 + m3 + m4
+        m[:2] # Results in m1 + m2, works
+        m[1:] # Should result in m2 + m3 + m4, does not work
+               # since m1 is part of all subexpressions.
+
+
+
+- Generally, all public methods have remained unchanged. The exception
+  is `.n_submodels` that used to be a method but now is a property.
```
Comment (nden) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  is ``.n_submodels`` that used to be a method but now is a property.
```

**Comment by nden** [comment]
One line missed test coverage but it's not related to documentation. Merging.

**Pull request merged**